formula/src/test/java/com/instacart/formula/FormulaRuntimeTest.kt

```diff
@@ -211,6 +214,113 @@ class FormulaRuntimeTest(val runtime: TestableRuntime, val name: String) {
         robot.test.output { assertThat(this).isEqualTo(3) }
     }
 
+    @Test fun `input change while running triggers root formula restart`() {
+        val terminationCallback = TestEventCallback<Int>()
+        var observer: TestFormulaObserver<Int, *, *>? = null
+        val root = object : StatelessFormula<Int, Int>() {
+            override fun key(input: Int): Any {
+                return input
+            }
+
+            override fun Snapshot<Int, Unit>.evaluate(): Evaluation<Int> {
+                return Evaluation(
+                    output = input,
+                    actions = context.actions {
+                        Action.onTerminate().onEvent {
+                            transition {
+                                terminationCallback.invoke(input)
+                            }
+                        }
+
+                        if (input == 0) {
+                            Action.onInit().onEvent {
+                                /**
+                                 * We call observer explicitly outside of effect block to ensure
+                                 * that input change happens while formula is running
+                                 */
+                                observer?.input(1)
+                                none()
+                            }
+                        }
+                    }
+                )
+            }
+        }
+
+        observer = runtime.test(root)
+        observer.input(0)
+        observer.output { assertThat(this).isEqualTo(1) }
+
+        // Check that termination was called
+        assertThat(terminationCallback.values()).containsExactly(0).inOrder()
```

Comment:
This is interesting behavior. Took me a few read-throughs to understand why this happens. I wonder if we should alert devs somehow when this occurs?

Reply:
Could you elaborate on the interesting part?

Reply:
Triggering onTerminate was surprising to me. I thought it would just discard the current evaluation and re-evaluate. 

Also I've seen a lot of people trigger side-effects in an onEvent block like this and I'm thinking if that side-effect triggers a new input this would similarly surprise people.

Reply:
It terminates because I have `key` defined for the formula
```kotlin
 override fun key(input: Int): Any {
      return input
  }
```

Reply:
Ah, I see now, that makes more sense.